Add `time` command and UI clock

byorgey opened the pull request:
Add `time` capability and related things:
- Add a new constant `time : cmd int`
- Make `clock` device provide `time` capability
- If the focused robot has a `clock` device installed (`base` has one by default in classic mode), display hexadecimal time in bottom left of the world panel.
- Add binary, octal, and hexadecimal integer literal syntax to the language.
- Make 16 ticks/s the default speed
    - The old default, 8 ticks/s, was annoyingly slow
    - Now you can simply type in `0x` followed by all the digits in the clock display, followed by a single trailing 0, to get the `time` in ticks

#402 discussed the possibility of having the `wait` command also require the `time` capability.  This breaks a number of tests which rely on the `wait` command.  I can fix the tests of course, but I'd like some feedback on whether requiring a `clock` in order to `wait` is a good idea.

Closes #402.
Closes #559.

TristanCacqueray commented on src/Swarm/TUI/View.hs:
```diff
@@ -233,6 +238,14 @@ drawGameUI s =
     Nothing -> id
+  -- Add clock display in bottom left of the world view if focused robot
+  -- has a clock installed
+  addClock = case s ^. gameState . to focusedRobot of
```
Could we also display `paused` when the game is not running, what ever the current focused robot is?

byorgey replied:
Great idea, done! :heavy_check_mark:

byorgey replied:
I also decided to show individual ticks in the clock display only when the game is paused, *i.e.* the decimal point and hex digit at the end of `0:00:03.a`.  That way you can see the tick count increase as you single-step the game.

xsebek commented:
> ... whether requiring a clock in order to wait is a good idea.

Technically getting something like `wait` manually requires arithmetic and recursion, so I would argue it should not be much harder than obtaining lambda and counter, but probably should not be free.

I think the clock requirement is sensible, though it will be annoying to add to challenges (like waiting for trees to grow). 

But hey, at least you can now do a speedrun of those challenges using the in-game clock! :smile:

xsebek commented:
Could the clock be on the top line instead? It's already crowded right above the REPL:
| ![image](https://user-images.githubusercontent.com/44544735/179844813-28766b29-86be-4826-9306-357e7392b322.png) |
|---|

AFAIK some OS put the clock top middle (macOS, Gnome) and some bottom right (Windows) by default. I would not look for the clock in the middle of the screen, but maybe some OS or program put it there.

xsebek commented on src/Swarm/Language/Parse.hs:
```diff
@@ -143,10 +143,20 @@ identifier = (lexeme . try) (p >>= check) <?> "variable name"
+  label "integer literal" $
+    lexeme $ do
+      n <-
+        try (string "0b" *> L.binary)
+          <|> try (string "0o" *> L.octal)
+          <|> try (string "0x" *> L.hexadecimal)
```
I think that will also need to be added to the editors highlighting. :sweat_smile: 

Should be an easy change to the `constant.numeric` match: `([0-9]*|0b[01]*|0o[0-8]*|0x\x*)`

xsebek reviewed (approved):
LGTM! :+1: It's definitely more fun to use step with the clock. :slightly_smiling_face:

byorgey commented:
> Could the clock be on the top line instead? It's already crowded right above the REPL:

Makes sense.  I moved the clock to the upper right, which seems like a natural place for a clock.  The cursor info used to be there, so I moved that down to the lower left: which actually seems more useful to me since I'm often copying numbers from the cursor info into a command I type at the REPL, so having it right next to the REPL helps.

byorgey commented:
> Technically getting something like `wait` manually requires arithmetic and recursion, so I would argue it should not be much harder than obtaining lambda and counter, but probably should not be free.

Yeah, that makes sense, but I see now what you mean about challenges --- the challenges themselves do not require `wait` (because a human can just wait long enough manually) but expressing a working `solution` requires `wait`.  It seems annoying and wrong to add a `clock` to the challenge just so we can write down a solution.

byorgey commented:
So I ended up adding a `clock` only in one case.  In a couple other cases I did some creative things to wait without requiring a `clock`.
